# Prosper Loan Data Exploration
## by Kirthi Kunhi Raman

## Preliminary Wrangling

> This document explores a dataset containing loan information of approximately 113938 loans provided to people for various reasons.

In [1]:
# import all packages and set plots to be embedded inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb

%matplotlib inline

In [2]:
# load in the dataset into a pandas dataframe, print statistics
loan = pd.read_csv('prosperLoanData.csv')

In [3]:
# high-level overview of data shape and composition
print(loan.shape)

(113937, 81)


In [4]:
print(loan.dtypes)

ListingKey                              object
ListingNumber                            int64
ListingCreationDate                     object
CreditGrade                             object
Term                                     int64
LoanStatus                              object
ClosedDate                              object
BorrowerAPR                            float64
BorrowerRate                           float64
LenderYield                            float64
EstimatedEffectiveYield                float64
EstimatedLoss                          float64
EstimatedReturn                        float64
ProsperRating (numeric)                float64
ProsperRating (Alpha)                   object
ProsperScore                           float64
ListingCategory (numeric)                int64
BorrowerState                           object
Occupation                              object
EmploymentStatus                        object
EmploymentStatusDuration               float64
IsBorrowerHom

In [5]:
print(loan.head(1))

                ListingKey  ListingNumber            ListingCreationDate  \
0  1021339766868145413AB3B         193129  2007-08-26 19:09:29.263000000   

  CreditGrade  Term LoanStatus           ClosedDate  BorrowerAPR  \
0           C    36  Completed  2009-08-14 00:00:00      0.16516   

   BorrowerRate  LenderYield    ...     LP_ServiceFees  LP_CollectionFees  \
0         0.158        0.138    ...            -133.18                0.0   

   LP_GrossPrincipalLoss  LP_NetPrincipalLoss LP_NonPrincipalRecoverypayments  \
0                    0.0                  0.0                             0.0   

   PercentFunded  Recommendations InvestmentFromFriendsCount  \
0            1.0                0                          0   

  InvestmentFromFriendsAmount Investors  
0                         0.0       258  

[1 rows x 81 columns]


In [6]:
# Remove columns that are not necessary in this analysis by dropping it
loandf = loan.copy()
loandf.drop(['ListingKey', 'ListingNumber', 'ListingCreationDate', 'CreditGrade', 'ClosedDate', 'LenderYield', 'EstimatedEffectiveYield', 'EstimatedLoss', 'EstimatedReturn', 'ProsperRating (numeric)', 'ProsperRating (Alpha)', 'ProsperScore', 'ListingCategory (numeric)', 'BorrowerState', 'Occupation', 'EmploymentStatusDuration', 'CurrentlyInGroup', 'GroupKey', 'DateCreditPulled', 'FirstRecordedCreditLine', 'CurrentCreditLines', 'TotalCreditLinespast7years', 'OpenRevolvingAccounts', 'OpenRevolvingMonthlyPayment', 'InquiriesLast6Months', 'TotalInquiries', 'CurrentDelinquencies', 'AmountDelinquent', 'DelinquenciesLast7Years', 'PublicRecordsLast10Years', 'PublicRecordsLast12Months', 'RevolvingCreditBalance', 'BankcardUtilization', 'AvailableBankcardCredit', 'TotalTrades', 'TradesNeverDelinquent (percentage)', 'TradesOpenedLast6Months', 'IncomeRange', 'IncomeVerifiable', 'LoanKey', 'TotalProsperLoans', 'TotalProsperPaymentsBilled', 'OnTimeProsperPayments', 'ProsperPaymentsLessThanOneMonthLate', 'ProsperPrincipalBorrowed', 'ProsperPrincipalOutstanding', 'ScorexChangeAtTimeOfListing', 'LoanCurrentDaysDelinquent', 'LoanFirstDefaultedCycleNumber', 'LoanMonthsSinceOrigination', 'LoanOriginationDate', 'LoanOriginationQuarter', 'MemberKey', 'LP_CustomerPayments', 'LP_CustomerPrincipalPayments', 'LP_InterestandFees', 'LP_ServiceFees', 'LP_CollectionFees', 'LP_GrossPrincipalLoss', 'LP_NetPrincipalLoss', 'LP_NonPrincipalRecoverypayments', 'PercentFunded', 'Recommendations', 'InvestmentFromFriendsCount', 'InvestmentFromFriendsAmount', 'Investors'], axis = 1, inplace = True) 

In [7]:
loandf.head()

Term LoanStatus  BorrowerAPR  BorrowerRate EmploymentStatus  \
0    36  Completed      0.16516        0.1580    Self-employed   
1    36    Current      0.12016        0.0920         Employed   
2    36  Completed      0.28269        0.2750    Not available   
3    36    Current      0.12528        0.0974         Employed   
4    36    Current      0.24614        0.2085         Employed   

   IsBorrowerHomeowner  CreditScoreRangeLower  CreditScoreRangeUpper  \
0                 True                  640.0                  659.0   
1                False                  680.0                  699.0   
2                False                  480.0                  499.0   
3                 True                  800.0                  819.0   
4                 True                  680.0                  699.0   

   OpenCreditLines  DebtToIncomeRatio  StatedMonthlyIncome  \
0              4.0               0.17          3083.333333   
1             14.0               0.18          6125.000000   
2              NaN               0.06          2083.333333   
3              5.0               0.15          2875.000000   
4             19.0               0.26          9583.333333   

   ProsperPaymentsOneMonthPlusLate  LoanNumber  LoanOriginalAmount  \
0                              NaN       19141                9425   
1                              NaN      134815               10000   
2                              NaN        6466                3001   
3                              NaN       77296               10000   
4                              0.0      102670               15000   

   MonthlyLoanPayment  
0              330.43  
1              318.93  
2              123.32  
3              321.45  
4              563.97

In [8]:
loandf.shape

(113937, 15)

In [9]:
# Determine duplicated rows
loandf[loandf.duplicated()]

Term LoanStatus  BorrowerAPR  BorrowerRate EmploymentStatus  \
9         36    Current      0.07620        0.0629         Employed   
999       60    Current      0.17710        0.1535         Employed   
2539      60    Current      0.14528        0.1224         Employed   
4942      36    Current      0.35356        0.3134         Employed   
5812      36    Current      0.24205        0.2045            Other   
5971      60    Current      0.18555        0.1620         Employed   
6385      36    Current      0.11563        0.0879         Employed   
7415      60    Current      0.24589        0.2209         Employed   
7538      36    Current      0.20524        0.1685         Employed   
7555      36    Current      0.12691        0.0990    Self-employed   
8683      36    Current      0.20524        0.1685         Employed   
8925      36    Current      0.22966        0.1920         Employed   
9102      36    Current      0.27817        0.2394         Employed   
9281      36    Current      0.34588        0.3059         Employed   
10036     36    Current      0.18990        0.1535         Employed   
10401     36    Current      0.29537        0.2566         Employed   
10467     36    Current      0.20268        0.1660         Employed   
11635     60    Current      0.15449        0.1314         Employed   
11889     36    Current      0.15223        0.1239         Employed   
11979     60    Current      0.14760        0.1249         Employed   
12379     60    Current      0.20040        0.1765         Employed   
12595     36    Current      0.18479        0.1485         Employed   
12694     60    Current      0.15844        0.1355         Employed   
12780     60    Current      0.15844        0.1355         Employed   
12956     36    Current      0.23540        0.1980         Employed   
13210     36    Current      0.14714        0.1189    Self-employed   
13322     36    Current      0.14853        0.1199         Employed   
13445     60    Current      0.26514        0.2394            Other   
14062     36    Current      0.19501        0.1585    Self-employed   
14071     36    Current      0.23233        0.1950         Employed   
...      ...        ...          ...           ...              ...   
111815    60    Current      0.20833        0.1840         Employed   
111839    60    Current      0.16969        0.1465         Employed   
111850    36    Current      0.11563        0.0879         Employed   
111860    36    Current      0.20933        0.1725         Employed   
111890    60    Current      0.17061        0.1474         Employed   
111991    36    Current      0.19144        0.1550         Employed   
111992    60    Current      0.17839        0.1550         Employed   
112216    36    Current      0.16732        0.1314         Employed   
112413    60    Current      0.21858        0.1940         Employed   
112495    36    Current      0.14714        0.1189         Employed   
112536    36    Current      0.14409        0.1159         Employed   
112603    60    Current      0.22549        0.2010         Employed   
112854    36    Current      0.21176        0.1745            Other   
112882    60    Current      0.21474        0.1905         Employed   
112919    60    Current      0.19221        0.1685            Other   
112924    36    Current      0.18275        0.1465         Employed   
112935    36    Current      0.22108        0.1840         Employed   
113025    60    Current      0.23293        0.2080            Other   
113136    36    Current      0.14714        0.1189         Employed   
113144    60    Current      0.14760        0.1249         Employed   
113218    36    Current      0.14039        0.1119         Employed   
113356    36    Current      0.26047        0.2225         Employed   
113386    36    Current      0.14243        0.1139         Employed   
113431    36    Current      0.20524        0.1685         Employed   
113468    60    Current      0.2

In [10]:
# Remove duplicated rows, keeping just the first out of 2 or more duplicated value
loandf.drop_duplicates(keep = 'first', inplace = True)

In [11]:
# Check if the 871 duplicated values are removed 
loandf.shape

(113066, 15)

In [12]:
# Check again if the duplicated values are removed by specifically checking if the duplicates of LoanNumber = 121268 are removed
loandf[loandf.LoanNumber == 121268]

Term LoanStatus  BorrowerAPR  BorrowerRate EmploymentStatus  \
8    36    Current       0.0762        0.0629         Employed   

   IsBorrowerHomeowner  CreditScoreRangeLower  CreditScoreRangeUpper  \
8                 True                  820.0                  839.0   

   OpenCreditLines  DebtToIncomeRatio  StatedMonthlyIncome  \
8             16.0               0.25          3333.333333   

   ProsperPaymentsOneMonthPlusLate  LoanNumber  LoanOriginalAmount  \
8                              NaN      121268               10000   

   MonthlyLoanPayment  
8              305.54

In [13]:
# Check for incomplete data in 'StatedMonthlyIncome'
loandf[loandf.StatedMonthlyIncome == 0]

Term              LoanStatus  BorrowerAPR  BorrowerRate  \
78        36              Chargedoff      0.18454        0.1700   
100       36              Chargedoff      0.35356        0.3134   
108       36               Completed      0.28320        0.2600   
125       36                 Current      0.35643        0.3199   
170       36               Completed      0.28851        0.2499   
186       36              Chargedoff      0.35797        0.3177   
332       36               Cancelled      0.25650        0.2375   
335       36               Completed      0.35797        0.3177   
688       36               Completed      0.30532        0.2699   
772       36               Completed      0.10491        0.0980   
917       36               Completed      0.15713        0.1500   
948       36               Completed      0.14770        0.1299   
954       36               Completed      0.27767        0.2700   
973       12               Completed      0.30461        0.2254   
988       36               Completed      0.38443        0.3220   
1014      36                 Current      0.24758        0.2099   
1157      36               Completed      0.12500        0.1180   
1314      36                 Current      0.31032        0.2712   
1346      36               Completed      0.37036        0.3500   
1478      36              Chargedoff      0.19730        0.1900   
1521      36              Chargedoff      0.14132        0.1200   
1529      60   Past Due (16-30 days)      0.27462        0.2489   
1557      36                 Current      0.35356        0.3134   
1773      60                 Current      0.24702        0.2220   
1790      36              Chargedoff      0.37453        0.3500   
1822      36               Completed      0.16991        0.1555   
1829      36                 Current      0.35132        0.3149   
1844      36               Completed      0.09436        0.0800   
1926      36              Chargedoff      0.30218        0.2900   
2064      36               Completed      0.15428        0.1400   
...      ...                     ...          ...           ...   
111707    36               Completed      0.35090        0.3108   
111934    36               Completed      0.13705        0.1300   
111974    36               Completed      0.28706        0.2638   
112052    36              Chargedoff      0.28974        0.2511   
112240    36               Completed      0.15713        0.1500   
112252    36               Completed      0.27306        0.2500   
112330    36                 Current      0.22742        0.1902   
112371    36                 Current      0.35356        0.3134   
112374    36              Chargedoff      0.16988        0.1627   
112383    36               Defaulted      0.15522        0.1481   
112398    36               Completed      0.08718        0.0700   
112491    36               Completed      0.22237        0.2000   
112670    36                 Current      0.35356        0.3134   
112685    12               Completed      0.31134        0.2320   
112784    60   Past Due (16-30 days)      0.22693        0.2024   
112821    36  Past Due (91-120 days)      0.20564        0.1764   
112917    36    Past Due (>120 days)      0.35797        0.3177   
112971    60              Chargedoff      0.24763        0.2249   
113197    36                 Current      0.35797        0.3177   
113409    36               Completed      0.05672        0.0500   
113417    36                 Current      0.35797        0.3177   
113485    36               Completed      0.18756        0.1730   
113608    60              Chargedoff      0.24252        0.2199   
113622    36              Chargedoff      0.34577        0.3058   
113682    60              Chargedoff      0.32680        0.2997   
113686    36               Completed      0.21588        0.2085   
113761    60                 Current      0.16151        0.1385   
113766    36                 Current      0.12427      

In [14]:
# If 'StatedMonthlyIncome' = 0, then we can drop these incomplete data since it does not make sense to predict loan repayment based on 0 income
index_names = loandf[loandf.StatedMonthlyIncome == 0 ].index
loandf.drop(index_names, inplace = True)

In [15]:
# Ensure that 'StatedMonthlyIncome' = 0 rows has been removed
loandf[loandf.StatedMonthlyIncome == 0]

Empty DataFrame
Columns: [Term, LoanStatus, BorrowerAPR, BorrowerRate, EmploymentStatus, IsBorrowerHomeowner, CreditScoreRangeLower, CreditScoreRangeUpper, OpenCreditLines, DebtToIncomeRatio, StatedMonthlyIncome, ProsperPaymentsOneMonthPlusLate, LoanNumber, LoanOriginalAmount, MonthlyLoanPayment]
Index: []

In [16]:
# Check for incomplete data in 'DebtToIncomeRatio'
loandf[loandf.DebtToIncomeRatio.isnull()]

Term  LoanStatus  BorrowerAPR  BorrowerRate EmploymentStatus  \
40        60     Current      0.19630        0.1725    Self-employed   
69        60     Current      0.13636        0.1139    Self-employed   
103       36  Chargedoff      0.26681        0.2287    Self-employed   
105       36  Chargedoff      0.30532        0.2699         Employed   
131       36   Completed      0.20623        0.1915        Part-time   
174       36     Current      0.09988        0.0864         Employed   
183       36     Current      0.13697        0.1089         Employed   
185       36   Defaulted      0.27119        0.2365    Self-employed   
204       60     Current      0.27246        0.2468         Employed   
209       36     Current      0.31790        0.2786         Employed   
216       36     Current      0.31790        0.2786    Self-employed   
218       36  Chargedoff      0.19236        0.1559    Self-employed   
255       36   Completed      0.26079        0.2379        Full-time   
291       60     Current      0.15868        0.1355    Self-employed   
301       36     Current      0.32538        0.2859    Self-employed   
303       36  Chargedoff      0.30532        0.2699    Self-employed   
307       36     Current      0.15223        0.1239    Self-employed   
313       36     Current      0.19144        0.1550    Self-employed   
315       36     Current      0.12081        0.0930    Self-employed   
330       36  Chargedoff      0.15377        0.1395    Self-employed   
331       36     Current      0.33973        0.2999    Self-employed   
337       36     Current      0.21025        0.1734    Self-employed   
351       36     Current      0.16678        0.1305    Self-employed   
354       36   Completed      0.37699        0.3400    Self-employed   
372       36  Chargedoff      0.37453        0.3500        Full-time   
390       36   Completed      0.30532        0.2699    Self-employed   
395       36   Completed      0.15512        0.1480    Self-employed   
429       36   Completed      0.23748        0.2300    Self-employed   
432       36   Completed      0.08001        0.0629    Self-employed   
473       36     Current      0.32446        0.2850    Self-employed   
...      ...         ...          ...           ...              ...   
113530    36     Current      0.35797        0.3177    Self-employed   
113541    36   Completed      0.37453        0.3500        Full-time   
113545    36     Current      0.25013        0.2124         Employed   
113550    36   Completed      0.23682        0.1990        Part-time   
113552    36   Completed      0.23631        0.1985        Full-time   
113555    36     Current      0.30532        0.2699    Self-employed   
113559    36     Current      0.35132        0.3149         Employed   
113600    36   Completed      0.28370        0.2452            Other   
113617    36   Completed      0.37453        0.3500    Self-employed   
113628    60   Completed      0.24589        0.2209    Self-employed   
113652    60     Current      0.16686        0.1435    Self-employed   
113654    36     Current      0.13799        0.1099    Self-employed   
113670    60     Current      0.24856        0.2235    Self-employed   
113672    36   Completed      0.35797        0.3177    Self-employed   
113684    60     Current      0.26877        0.2432    Self-employed   
113689    12   Completed      0.29901        0.2199    Self-employed   
113704    36   Completed      0.05104        0.0500        Full-time   
113711    36   Completed      0.25278        0.2300    Self-employed   
113725    36   Completed      0.11323        0.0840        Part-time   
113744    36     Current      0.12691        0.0990    Self-employed   
113792    36   Completed      0.26681        0.2287    Self-employed   
113836    36   Completed      0.23794        0.2001    Self-employed   
113848    36   Completed      0.30709        0.2875    Not available   
113857    36     Current      0.28974        0.2

In [17]:
# Predicting the loan outcome status and interest values with missing DebtToIncomeRatio value is not practical/hard.
# Remove rows with no/missing (NaN) DebtToIncomeRatio value
index_names = loandf[loandf.DebtToIncomeRatio.isnull()].index
loandf.drop(index_names, inplace = True)

In [18]:
# Ensure that 'DebtToIncomeRatio' = NaN rows has been removed
loandf[loandf.DebtToIncomeRatio.isnull()]

Empty DataFrame
Columns: [Term, LoanStatus, BorrowerAPR, BorrowerRate, EmploymentStatus, IsBorrowerHomeowner, CreditScoreRangeLower, CreditScoreRangeUpper, OpenCreditLines, DebtToIncomeRatio, StatedMonthlyIncome, ProsperPaymentsOneMonthPlusLate, LoanNumber, LoanOriginalAmount, MonthlyLoanPayment]
Index: []

In [19]:
# Check if OpenCreditLines column has NaN value
loandf[loandf.OpenCreditLines.isnull()]

Term  LoanStatus  BorrowerAPR  BorrowerRate EmploymentStatus  \
2         36   Completed      0.28269        0.2750    Not available   
17        36  Chargedoff      0.13202        0.1250    Not available   
39        36   Completed      0.15211        0.1450              NaN   
42        36  Chargedoff      0.30300        0.2900    Not available   
63        36   Completed      0.29776        0.2900    Not available   
76        36   Completed      0.25861        0.2500    Not available   
90        36   Completed      0.19178        0.1845    Not available   
92        36   Completed      0.08292        0.0755              NaN   
122       36   Completed      0.08718        0.0700    Not available   
128       36   Defaulted      0.18824        0.1700    Not available   
146       36   Completed      0.09939        0.0925              NaN   
162       36   Completed      0.08483        0.0780              NaN   
172       36  Chargedoff      0.24753        0.2400    Not available   
178       36   Completed      0.24856        0.2400    Not available   
184       36   Defaulted      0.26590        0.2500    Not available   
191       36   Defaulted      0.30781        0.3000    Not available   
193       36   Defaulted      0.23937        0.2300              NaN   
206       36   Defaulted      0.27767        0.2700              NaN   
215       36   Defaulted      0.18726        0.1800    Not available   
217       36   Defaulted      0.29776        0.2900    Not available   
247       36   Completed      0.19582        0.1775              NaN   
249       36  Chargedoff      0.29776        0.2900    Not available   
250       36   Completed      0.08857        0.0800              NaN   
251       36   Defaulted      0.16647        0.1593    Not available   
252       36  Chargedoff      0.29925        0.2900    Not available   
275       36   Completed      0.20896        0.1905    Not available   
298       36   Completed      0.19128        0.1840    Not available   
310       36   Completed      0.12724        0.1175              NaN   
325       36   Completed      0.13453        0.1275              NaN   
336       36   Defaulted      0.17904        0.1700    Not available   
...      ...         ...          ...           ...              ...   
113452    36   Completed      0.09809        0.0912    Not available   
113453    36   Completed      0.08784        0.0810    Not available   
113454    36   Defaulted      0.11596        0.1090    Not available   
113459    36  Chargedoff      0.25707        0.2495    Not available   
113467    36   Completed      0.16215        0.1550    Not available   
113494    36   Completed      0.23748        0.2300              NaN   
113496    36   Completed      0.17722        0.1700    Not available   
113515    36  Chargedoff      0.22931        0.2200    Not available   
113531    36   Defaulted      0.21857        0.2000    Not available   
113591    36   Completed      0.15462        0.1475              NaN   
113626    36   Completed      0.17722        0.1700              NaN   
113629    36  Chargedoff      0.29776        0.2900    Not available   
113630    36   Completed      0.20735        0.2000    Not available   
113635    36   Defaulted      0.29776        0.2900    Not available   
113663    36  Chargedoff      0.29525        0.2875              NaN   
113664    36   Completed      0.21689        0.2095    Not available   
113712    36   Defaulted      0.24502        0.2375              NaN   
113714    36  Chargedoff      0.28887        0.2695    Not available   
113728    36   Completed      0.16014        0.1530    Not available   
113732    36  Chargedoff      0.29776        0.2900    Not available   
113735    36  Chargedoff      0.22141        0.2140    Not available   
113736    36   Completed      0.16185        0.1547    Not available   
113768    36   Completed      0.17722        0.1700    Not available   
113797    36   Defaulted      0.11696        0.1

In [20]:
# In OpenCreditLines column,replace NaN value with 0
loandf.OpenCreditLines = loandf.OpenCreditLines.fillna(0)

In [21]:
# Ensure OpenCreditLines column has No NaN value
loandf[loandf.OpenCreditLines.isnull()]

Empty DataFrame
Columns: [Term, LoanStatus, BorrowerAPR, BorrowerRate, EmploymentStatus, IsBorrowerHomeowner, CreditScoreRangeLower, CreditScoreRangeUpper, OpenCreditLines, DebtToIncomeRatio, StatedMonthlyIncome, ProsperPaymentsOneMonthPlusLate, LoanNumber, LoanOriginalAmount, MonthlyLoanPayment]
Index: []

In [22]:
# Check if NaN value exists in CreditScoreRangeLower column
loandf[loandf.CreditScoreRangeLower.isnull()]

Term  LoanStatus  BorrowerAPR  BorrowerRate EmploymentStatus  \
206       36   Defaulted      0.27767        0.2700              NaN   
387       36   Completed      0.09337        0.0865              NaN   
698       36   Completed      0.07680        0.0700              NaN   
1023      36   Completed      0.08684        0.0800              NaN   
1126      36   Completed      0.20735        0.2000              NaN   
1148      36   Completed      0.10692        0.1000              NaN   
1907      36   Completed      0.11094        0.1040              NaN   
1932      36   Defaulted      0.18658        0.1775              NaN   
2029      36   Completed      0.13956        0.1325              NaN   
2040      36   Completed      0.10133        0.0840              NaN   
2106      36   Completed      0.35604        0.3480              NaN   
2204      36   Completed      0.24502        0.2375              NaN   
2347      36   Completed      0.09588        0.0890              NaN   
2543      36   Completed          NaN        0.1000              NaN   
2872      36   Completed      0.11696        0.1100              NaN   
3004      36   Completed      0.09437        0.0875              NaN   
3008      36   Defaulted      0.15713        0.1500              NaN   
3191      36   Completed      0.36559        0.3575              NaN   
3234      36   Completed      0.23748        0.2300              NaN   
3329      36   Defaulted      0.10090        0.0940              NaN   
3435      36   Defaulted      0.15713        0.1500              NaN   
3578      36   Completed      0.12700        0.1200              NaN   
3695      36   Completed      0.08684        0.0800              NaN   
3881      36   Completed      0.07830        0.0612              NaN   
4271      36   Defaulted      0.10692        0.1000              NaN   
4823      36   Completed      0.12700        0.1200              NaN   
4910      36   Completed      0.18726        0.1800              NaN   
4994      36  Chargedoff      0.20684        0.1995              NaN   
5367      36   Completed      0.13705        0.1300              NaN   
5871      36  Chargedoff      0.14935        0.1400              NaN   
...      ...         ...          ...           ...              ...   
106735    36   Completed          NaN        0.0595              NaN   
107126    36   Completed      0.08935        0.0825              NaN   
107726    36   Completed      0.13052        0.1235              NaN   
108137    36  Chargedoff      0.22744        0.2200              NaN   
108436    36   Completed      0.25255        0.2450              NaN   
108643    36   Defaulted      0.20735        0.2000              NaN   
109027    36   Completed          NaN        0.0600              NaN   
109622    36   Completed      0.28771        0.2800              NaN   
109832    36   Completed      0.15713        0.1500              NaN   
110012    36   Completed      0.10441        0.0975              NaN   
110108    36   Completed      0.15311        0.1460              NaN   
110207    36   Defaulted      0.32791        0.3200              NaN   
110279    36   Completed      0.16466        0.1575              NaN   
110396    36   Defaulted      0.15713        0.1500              NaN   
110537    36   Completed      0.08383        0.0770              NaN   
110964    36   Completed      0.20846        0.1900              NaN   
110988    36   Completed      0.08182        0.0750              NaN   
111184    36   Completed      0.24753        0.2400              NaN   
111267    36   Completed      0.22744        0.2200              NaN   
111405    36  Chargedoff      0.18964        0.1775              NaN   
111607    36   Completed          NaN        0.0650              NaN   
111853    36   Completed      0.09728        0.0800              NaN   
112198    36   Completed      0.25650        0.2375              NaN   
112387    36   Completed      0.11144        0.1

In [23]:
# remove rows with CreditScoreRangeLower = NaN, since it does not help in predicting interest rate when credit score is missing.
index_names = loandf[loandf.CreditScoreRangeLower.isnull()].index
loandf.drop(index_names, inplace = True)

In [24]:
# Ensure that all rows with CreditScoreRangeLower = NaN has been removed
loandf[loandf.CreditScoreRangeLower.isnull()]

Empty DataFrame
Columns: [Term, LoanStatus, BorrowerAPR, BorrowerRate, EmploymentStatus, IsBorrowerHomeowner, CreditScoreRangeLower, CreditScoreRangeUpper, OpenCreditLines, DebtToIncomeRatio, StatedMonthlyIncome, ProsperPaymentsOneMonthPlusLate, LoanNumber, LoanOriginalAmount, MonthlyLoanPayment]
Index: []

In [25]:
# Check if rows with CreditScoreRangeUpper = NaN exists
loandf[loandf.CreditScoreRangeUpper.isnull()]

Empty DataFrame
Columns: [Term, LoanStatus, BorrowerAPR, BorrowerRate, EmploymentStatus, IsBorrowerHomeowner, CreditScoreRangeLower, CreditScoreRangeUpper, OpenCreditLines, DebtToIncomeRatio, StatedMonthlyIncome, ProsperPaymentsOneMonthPlusLate, LoanNumber, LoanOriginalAmount, MonthlyLoanPayment]
Index: []

In [26]:
# Check if NaN value exists in ProsperPaymentsOneMonthPlusLate column
loandf[loandf.ProsperPaymentsOneMonthPlusLate.isnull()]

Term            LoanStatus  BorrowerAPR  BorrowerRate  \
0         36             Completed      0.16516        0.1580   
1         36               Current      0.12016        0.0920   
2         36             Completed      0.28269        0.2750   
3         36               Current      0.12528        0.0974   
5         60               Current      0.15425        0.1314   
6         36               Current      0.31032        0.2712   
7         36               Current      0.23939        0.2019   
8         36               Current      0.07620        0.0629   
10        60               Current      0.27462        0.2489   
11        36             Completed      0.15033        0.1325   
12        36  Past Due (1-15 days)      0.17969        0.1435   
13        36               Current      0.13138        0.1034   
14        60               Current      0.11695        0.0949   
15        36             Defaulted      0.35797        0.3177   
17        36            Chargedoff      0.13202        0.1250   
18        36               Current      0.12528        0.0974   
20        36               Current      0.16732        0.1314   
21        36             Completed      0.21488        0.2075   
22        36               Current      0.35356        0.3134   
23        36             Defaulted      0.28032        0.2419   
24        36               Current      0.19859        0.1620   
25        36               Current      0.30182        0.2629   
26        60             Completed      0.30748        0.2809   
27        36             Completed      0.11296        0.0920   
28        36               Current      0.20268        0.1660   
29        36               Current      0.15223        0.1239   
30        36             Completed      0.12782        0.0999   
31        36               Current      0.35797        0.3177   
32        36               Current      0.16324        0.1274   
34        36               Current      0.22966        0.1920   
...      ...                   ...          ...           ...   
113890    36             Completed      0.15462        0.1475   
113891    36             Completed      0.19543        0.1808   
113892    36               Current      0.15223        0.1239   
113896    36            Chargedoff      0.32327        0.2995   
113897    36             Completed      0.12550        0.1185   
113898    36             Completed      0.16005        0.1385   
113900    36               Current      0.16678        0.1305   
113901    60               Current      0.15016        0.1274   
113904    36            Chargedoff      0.24753        0.2400   
113905    36            Chargedoff      0.35797        0.3177   
113906    36             Completed      0.07479        0.0680   
113907    36               Current      0.35356        0.3134   
113909    60               Current      0.27554        0.2498   
113910    36               Current      0.32446        0.2850   
113911    36               Current      0.25330        0.2155   
113913    36             Completed      0.12410        0.1030   
113914    36             Defaulted      0.25757        0.2500   
113916    36               Current      0.33286        0.2932   
113917    36               Current      0.32446        0.2850   
113918    36             Defaulted      0.12201        0.1080   
113919    36               Current      0.27285        0.2346   
113922    36             Completed      0.15094        0.1295   
113923    36             Completed      0.22378        0.2089   
113925    36               Current      0.31032        0.2712   
113926    36             Defaulted      0.29776        0.2900   
113929    36             Completed      0.30285        0.2639   
113931    60               Current      0.15016        0.1274   
113932    36               Current      0.22354        0.1864   
113934    60               Current      0.23984        0.2150   
113936    36               Current      0.1

In [27]:
# In ProsperPaymentsOneMonthPlusLate column,replace NaN value with 0
loandf.ProsperPaymentsOneMonthPlusLate = loandf.ProsperPaymentsOneMonthPlusLate.fillna(0)

In [28]:
# Check if all NaN values are replaced with 0
loandf[loandf.ProsperPaymentsOneMonthPlusLate.isnull()]

Empty DataFrame
Columns: [Term, LoanStatus, BorrowerAPR, BorrowerRate, EmploymentStatus, IsBorrowerHomeowner, CreditScoreRangeLower, CreditScoreRangeUpper, OpenCreditLines, DebtToIncomeRatio, StatedMonthlyIncome, ProsperPaymentsOneMonthPlusLate, LoanNumber, LoanOriginalAmount, MonthlyLoanPayment]
Index: []

In [29]:
# convert ProsperPaymentsOneMonthPlusLate column values to int values
loandf.ProsperPaymentsOneMonthPlusLate = loandf.ProsperPaymentsOneMonthPlusLate.astype(int)

In [30]:
# Rows and columns in the  dataset
loandf.shape

(103998, 15)

In [31]:
loandf.head()

Term LoanStatus  BorrowerAPR  BorrowerRate EmploymentStatus  \
0    36  Completed      0.16516        0.1580    Self-employed   
1    36    Current      0.12016        0.0920         Employed   
2    36  Completed      0.28269        0.2750    Not available   
3    36    Current      0.12528        0.0974         Employed   
4    36    Current      0.24614        0.2085         Employed   

   IsBorrowerHomeowner  CreditScoreRangeLower  CreditScoreRangeUpper  \
0                 True                  640.0                  659.0   
1                False                  680.0                  699.0   
2                False                  480.0                  499.0   
3                 True                  800.0                  819.0   
4                 True                  680.0                  699.0   

   OpenCreditLines  DebtToIncomeRatio  StatedMonthlyIncome  \
0              4.0               0.17          3083.333333   
1             14.0               0.18          6125.000000   
2              0.0               0.06          2083.333333   
3              5.0               0.15          2875.000000   
4             19.0               0.26          9583.333333   

   ProsperPaymentsOneMonthPlusLate  LoanNumber  LoanOriginalAmount  \
0                                0       19141                9425   
1                                0      134815               10000   
2                                0        6466                3001   
3                                0       77296               10000   
4                                0      102670               15000   

   MonthlyLoanPayment  
0              330.43  
1              318.93  
2              123.32  
3              321.45  
4              563.97

### What is the structure of your dataset?

> There are 113937 loan cases in the original dataset with 81 features (ListingKey, ListingNumber, ListingCreationDate, CreditGrade, Term, LoanStatus, ClosedDate, BorrowerAPR, etc). Most variables are numeric in nature, but the there are some nominal data too.<br>
> After data wrangling, now there are 103998 loan cases with 15 features essential for the analysis.

### What is/are the main feature(s) of interest in your dataset?

> I'm most interested in figuring out:
1. What factors affect a loan's outcome status?
2. What affects the borrower's APR or interest rate?
3. Are there differences between loans depending on how large the original loan amount was?

### What features in the dataset do you think will help support your investigation into your feature(s) of interest?

> I expect that factors such as EmploymentStatus, ProsperPaymentsOneMonthPlusLate, StatedMonthlyIncome, DebtToIncomeRatio and OpenCreditLines will have strongest effect on loan's outcome status.<br>
> Factors like LoanOriginalAmount, Term, CreditScoreRangeLower and CreditScoreRangeUpper will have affect borrower's interest rate. <br>
> Difference between loans based on large original amount can be investigated using factors like StatedMonthlyIncome, CreditScoreRangeLower and DebtToIncomeRatio.
